### PR TITLE
feat(celldescriptor): add CellDescriptor

### PR DESCRIPTION
**Pull request by aleksei-bobrov**

### CODE REVIEW

File: DataSource/DataSourceCellDescriptor.swift [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+	{
+		self.init(reuseIdentifier, prototypeSource) { $1 is Item }
+	}
+}
+
+extension CollectionViewDataSource {
+	func configure(_ collectionView: UICollectionView, using cellDescriptors: [CellDescriptor]) {
+		self.reuseIdentifierForItem = { indexPath, item in
+			for descriptor in cellDescriptors {
+				if descriptor.isMatching(indexPath, item) {
```
Comment (stephanecopin) [inline DataSource/DataSourceCellDescriptor.swift]: We can use `first(where:)` here, so something like:
```swift
return cellDescriptors.first { $0.isMatching(indexPath, item) }?.reuseIdentifier ?? fatalError()
```

File: DataSource/DataSourceCellDescriptor.swift [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+		self.collectionView = collectionView
+	}
+}
+
+extension TableViewDataSource {
+	func configure(_ tableView: UITableView, using cellDescriptors: [CellDescriptor]) {
+		self.reuseIdentifierForItem = { indexPath, item in
+			for descriptor in cellDescriptors {
+				if descriptor.isMatching(indexPath, item) {
+					return descriptor.reuseIdentifier
```
Comment (stephanecopin) [inline DataSource/DataSourceCellDescriptor.swift]: Same

File: DataSource/DataSourceCellDescriptor.swift [hunk 3]
```diff
@@ -0,0 +1,81 @@
+import UIKit
+
+enum CellPrototypeSource {
```
Comment (stephanecopin) [inline DataSource/DataSourceCellDescriptor.swift]: I feel like this should be nested within `CellDescriptor`, with the name `PrototypeSource`

Comment (aleksei-bobrov) [reply DataSource/DataSourceCellDescriptor.swift]: 👍

File: DataSource/DataSourceCellDescriptor.swift (hunk 1 above)
Comment (aleksei-bobrov) [reply DataSource/DataSourceCellDescriptor.swift]: Looks better ✅

File: DataSource/DataSourceCellDescriptor.swift (hunk 2 above)
Comment (aleksei-bobrov) [reply DataSource/DataSourceCellDescriptor.swift]: Looks better ✅

**Comment by notbenoit** [comment]
👍

File: DataSource/DataSourceCellDescriptor.swift [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+		case storyboard
+		case nib(UINib)
+		case `class`(AnyObject.Type)
+	}
+}
+
+extension CollectionViewDataSource {
+	func configure(_ collectionView: UICollectionView, using cellDescriptors: [CellDescriptor]) {
+		self.reuseIdentifierForItem = { indexPath, item in
+			guard let reuseIdentifier = cellDescriptors.first(where: { $0.isMatching(indexPath,item) })?.reuseIdentifier else {
```
Comment (heymansmile) [inline DataSource/DataSourceCellDescriptor.swift]: `$0.isMatching(indexPath,item)` space after the comma please.

File: DataSource/DataSourceCellDescriptor.swift [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+		_ reuseIdentifier: String,
+		_ itemType: Item.Type,
+		_ prototypeSource: PrototypeSource = .storyboard)
+	{
+		self.init(reuseIdentifier, prototypeSource) { $1 is Item }
+	}
+}
+
+extension CellDescriptor {
+	enum PrototypeSource {
```
Comment (heymansmile) [inline DataSource/DataSourceCellDescriptor.swift]: Why in the extension btw?

Comment (aleksei-bobrov) [reply DataSource/DataSourceCellDescriptor.swift]: In order to avoid bloating up the structure with additional definitions

Comment (heymansmile) [reply DataSource/DataSourceCellDescriptor.swift]: Makes sense 👍

**Comment by heymansmile** [comment]
Welcome home `CellDescriptor ` !!!

**Pull request merged**